### PR TITLE
Kronecker prod

### PR DESCRIPTION
**Pull request by izmailovpavel**
Operations for Kronecker products of square matrices (tt-ranks=1, tt-cores are square)

### CODE REVIEW

File: kronecker.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    raise ValueError('The argument should be a Kronecker product (tt-ranks should be 1)')
+
+  cores = kron_a.tt_cores
+  det, pows = 1, 1
+  for core_idx in range(kron_a.ndims()):
+    core = kron_a.tt_cores[core_idx]
+    if core.get_shape()[1] != core.get_shape()[2]:
+      raise ValueError('The argument should be a Kronecker product of square matrices' 
+                      '(tt-cores must be square)')
+    pows *= core.get_shape()[1].value
```
Comment (Bihaqo) [inline kronecker.py]: I think it's better to use tf.reduce_prod(t3f.get_raw_shape(kron_a)[1]) instead of a loop here

File: kronecker.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+def slog_determinant(kron_a):
+  """Computes the sign and (natural) logarithm of the determinant of 
+  the given matrix, factorized intoa Kronecker product of square matrices.
+
+  Args:
+    kron_a: `TensorTrain` object containing a matrix of size N x N, 
+    factorized into a Kronecker product of square matrices (all 
+    tt-ranks are 1 and all tt-cores are square).  
+  Returns:
+    Number, the log-determinant of the given matrix
```
Comment (Bihaqo) [inline kronecker.py]: Not true

File: kronecker.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+  if not _is_kron(kron_a):
+    raise ValueError('The argument should be a Kronecker product (tt-ranks should be 1)')
+  
+  pows = 1
+  for core_idx in range(kron_a.ndims()):
+    core = kron_a.tt_cores[core_idx]
+    if core.get_shape()[1] != core.get_shape()[2]:
+      raise ValueError('The argument should be a Kronecker product of square matrices' 
+                      '(tt-cores must be square)')
+    pows *= core.get_shape()[1].value
```
Comment (Bihaqo) [inline kronecker.py]: tf.reduce_sum(tf.log(ops.get_raw_shape(kron_a)[0]))

File: kronecker.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+                                                          
+  logdet = 0
+  det_sign = 1
+  for core_idx in range(kron_a.ndims()):
+    core = kron_a.tt_cores[core_idx]
+    core_det = tf.matrix_determinant(core[0, :, :, 0])
+    core_abs_det = tf.abs(core_det)
```
Comment (Bihaqo) [inline kronecker.py]: Would be cool to do liky in numpy:
If the determinant is zero, then sign will be 0 and logdet will be
-Inf. In all cases, the determinant is equal to sign * np.exp(logdet).

At least, add a TODO

File: kronecker.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+  for core_idx in range(kron_a.ndims()):
+    core = kron_a.tt_cores[core_idx]
+    if core.get_shape()[1] != core.get_shape()[2]:
+      raise ValueError('The argument should be a Kronecker product of square matrices' 
+                      '(tt-cores must be square)')
+ 
+    core_inv = tf.matrix_inverse(core[0, :, :, 0])
+    inv_cores.append(tf.expand_dims(tf.expand_dims(core_inv, 0), -1))
+
+  res_ranks = tf.TensorShape([1] * (kron_a.ndims() + 1))
```
Comment (Bihaqo) [inline kronecker.py]: kron_a.get_tt_ranks()

File: kronecker.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+  cho_cores = []
+  for core_idx in range(kron_a.ndims()):
+    core = kron_a.tt_cores[core_idx]
+    if core.get_shape()[1] != core.get_shape()[2]:
+      raise ValueError('The argument should be a Kronecker product of square matrices' 
+                      '(tt-cores must be square)')
+    core_cho = tf.cholesky(core[0, :, :, 0])
+    cho_cores.append(tf.expand_dims(tf.expand_dims(core_cho, 0), -1))
+
+  res_ranks = tf.TensorShape([1] * (kron_a.ndims() + 1))
```
Comment (Bihaqo) [inline kronecker.py]: kron_a.get_tt_ranks()

File: kronecker.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+  """Returns True if the argument is a Kronecker product matrix.
+
+  Args:
+    tt_a: `TensorTrain` object
+
+  Returns:
+    bool
+  """
+  if tt_a.is_tt_matrix():
+    return max(tt_a.get_tt_ranks()) == 1
```
Comment (Bihaqo) [inline kronecker.py]: May make sense to implement TensorTrain.get_max_tt_rank()

File: kronecker_test.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+  def testSlogDet(self):
+    """
+    Tests the slog_determinant function
+    """
+    tf.set_random_seed(5) # negative derminant
+    kron_neg = t3f.get_variable('kron_neg', initializer=t3f.random_matrix(((2, 3), (2, 3)), tt_rank=1))
+  
+    tf.set_random_seed(1) # positive determinant
+    kron_pos = t3f.get_variable('kron_pos', initializer=t3f.random_matrix(((2, 3), (2, 3)), tt_rank=1))
```
Comment (Bihaqo) [inline kronecker_test.py]: Add todo to use kron and -1 * kron when we implement __mul__ operation. The current version is dangerous -- random seed is not reliable across platforms, devices and future versions.

File: kronecker.py [hunk 9]
```diff
@@ -4,9 +4,8 @@
 from tensor_train import TensorTrain
 
 def determinant(kron_a):
-  """Computes the determinant of a given matrix, factorized into
-  a Kronecker product of square matrices.
-  
+  """Computes the determinant of a given Kronecker-factorized matrix
```
Comment (Bihaqo) [inline kronecker.py]: use dot at the end of the docstring sentence

File: kronecker.py [hunk 10]
```diff
@@ -22,15 +21,16 @@ def determinant(kron_a):
     or the tt-ranks are not 1
   """
   if not _is_kron(kron_a):
-    raise ValueError('The argument should be a Kronecker product (tt-ranks should be 1)')
+    raise ValueError('The argument should be a Kronecker product (tt-ranks'
```
Comment (Bihaqo) [inline kronecker.py]: Add a space at the end, or it would be '(tt-ranksshould be 1)'

File: kronecker_test.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+import t3f
+import tensor_train
+import kronecker as kr
+import ops
+
+class KroneckerTest(tf.test.TestCase):
+
+  def testIsKronNonKron(self):
+    """
```
Comment (izmailovpavel) [inline kronecker_test.py]: Done

File: kronecker_test.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    Tests _is_kron on a Kronecker matrix
+    """
+    kron_mat = t3f.get_variable('kron_mat', initializer=t3f.random_matrix(((2, 3), (3, 2)), tt_rank=1))
+    self.assertTrue(kr._is_kron(kron_mat))
+
+  def testDet(self):
+    """
+    Tests the determinant function
+    """
+    tf.set_random_seed(5)
```
Comment (izmailovpavel) [inline kronecker_test.py]: Done

File: kronecker_test.py (hunk 8 above)
Comment (izmailovpavel) [inline kronecker_test.py]: Done

File: kronecker.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
+import numpy as np
+import tensorflow as tf
+
+from tensor_train import TensorTrain
+
+def determinant(kron_a):
+  """Computes the determinant of a given matrix, factorized into
```
Comment (izmailovpavel) [inline kronecker.py]: Done

File: kronecker.py [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+    or the tt-ranks are not 1
+  """
+  if not _is_kron(kron_a):
+    raise ValueError('The argument should be a Kronecker product (tt-ranks should be 1)')
+
+  cores = kron_a.tt_cores
+  det, pows = 1, 1
+  for core_idx in range(kron_a.ndims()):
+    core = kron_a.tt_cores[core_idx]
+    if core.get_shape()[1] != core.get_shape()[2]:
```
Comment (izmailovpavel) [inline kronecker.py]: Done

File: kronecker.py (hunk 1 above)
Comment (izmailovpavel) [inline kronecker.py]: Done

File: kronecker.py (hunk 2 above)
Comment (izmailovpavel) [inline kronecker.py]: Done

File: kronecker.py (hunk 3 above)
Comment (izmailovpavel) [inline kronecker.py]: We don't need to compute the log of this product here, so I used what you suggested in the previous comment here

File: kronecker.py (hunk 4 above)
Comment (izmailovpavel) [inline kronecker.py]: It actually does behave like that already

File: kronecker.py (hunk 5 above)
Comment (izmailovpavel) [inline kronecker.py]: Done

File: kronecker.py (hunk 9 above)
Comment (izmailovpavel) [inline kronecker.py]: Done

File: kronecker.py (hunk 10 above)
Comment (izmailovpavel) [inline kronecker.py]: Done

File: kronecker.py (hunk 6 above)
Comment (izmailovpavel) [inline kronecker.py]: Done
